google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/StreamingSubscriberConnection.java

```diff
@@ -321,50 +417,473 @@ private boolean isAlive() {
+        });
+
+    sendAcks(ackIdMessageFutures);
+  }
 
-          @Override
-          public void onFailure(Throwable t) {
-            ackOperationsWaiter.incrementPendingCount(-1);
-            Level level = isAlive() ? Level.WARNING : Level.FINER;
-            logger.log(level, "failed to send operations", t);
+  private class RetryModacksFailedModacks {
+    private List<ModackWithMessageFuture> retryModackWithMessageFutures;
+    private Set<String> failedAckIds;
+
+    RetryModacksFailedModacks(
+        List<ModackWithMessageFuture> retryModackWithMessageFutures, Set<String> failedAckIds) {
+      this.retryModackWithMessageFutures = retryModackWithMessageFutures;
+      this.failedAckIds = failedAckIds;
+    }
+
+    public List<ModackWithMessageFuture> getRetryModackWithMessageFutures() {
+      return retryModackWithMessageFutures;
+    }
+
+    public Set<String> getFailedAckIds() {
+      return failedAckIds;
+    }
+  }
+
+  private Map<String, String> getMetadataMapFromThrowable(Throwable t) {
+    // This converts a Throwable (from a "OK" grpc response) to a map of metadata
+    // will be of the format:
+    // {
+    //    "ACK-ID-1": "PERMANENT_*",
+    //    "ACK-ID-2": "TRANSIENT_*"
+    // }
+    com.google.rpc.Status status = StatusProto.fromThrowable(t);
+    Map<String, String> metadataMap = new HashMap<>();
+    if (status != null) {
+      for (Any any : status.getDetailsList()) {
+        if (any.is(ErrorInfo.class)) {
+          try {
+            ErrorInfo errorInfo = any.unpack(ErrorInfo.class);
+            metadataMap = errorInfo.getMetadataMap();
+          } catch (Throwable throwable) {
           }
-        };
+        }
+      }
+    }
+    return metadataMap;
+  }
+
+  private ApiFutureCallback<Empty> getLoggingCallback(
+      SettableApiFuture<Map<String, String>> responseFuture) {
+    return new ApiFutureCallback<Empty>() {
+      @Override
+      public void onSuccess(Empty empty) {
+        ackOperationsWaiter.incrementPendingCount(-1);
+        responseFuture.set(new HashMap<>());
+      }
 
+      @Override
+      public void onFailure(Throwable t) {
+        ackOperationsWaiter.incrementPendingCount(-1);
+        Map<String, String> metadataMap = getMetadataMapFromThrowable(t);
+        if (!metadataMap.isEmpty()) {
+          responseFuture.set(metadataMap);
+        }
+        Level level = isAlive() ? Level.WARNING : Level.FINER;
+        logger.log(level, "failed to send operations", t);
+      }
+    };
+  }
+
+  private ApiFutureCallback<Empty> getLoggingCallback() {
+    return new ApiFutureCallback<Empty>() {
+      @Override
+      public void onSuccess(Empty empty) {
+        ackOperationsWaiter.incrementPendingCount(-1);
+      }
+
+      @Override
+      public void onFailure(Throwable t) {
+        ackOperationsWaiter.incrementPendingCount(-1);
+        Level level = isAlive() ? Level.WARNING : Level.FINER;
+        logger.log(level, "failed to send operations", t);
+      }
+    };
```

Comment:
nit. We can probably use the same logging callback, and pass in a java.util.Optional responseFuture, and only do exactly once stuff if the response future is provided.